dgisim/src/dices.py

```diff
@@ -1,29 +1,62 @@
 from __future__ import annotations
-
 import random
+from collections import Counter
+from functools import lru_cache
+from typing import Any, Optional, Iterator, Iterable
+from typing_extensions import Self, override, TYPE_CHECKING
 from collections import defaultdict
-from typing import Optional, Iterator, Iterable, Union, List, Set, Tuple, Dict
+from functools import cache
+from heapq import heappop, heapify
+
+from typing_extensions import TYPE_CHECKING, Self
 
-from typing_extensions import Self
+if TYPE_CHECKING:
+    from .state.game_state import GameState
+    from .state.player_state import PlayerState
 
-from dgisim.src.element.element import Element
+from dgisim.src.element import Element
 from dgisim.src.helper.hashable_dict import HashableDict
-from dgisim.src.helper.level_print import level_print
-from dgisim.src.helper.quality_of_life import BIG_INT
+from dgisim.src.helper.quality_of_life import BIG_INT, case_val
 
-ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY: List[Element] = [Element.PYRO,
-                                                         Element.HYDRO,
-                                                         Element.ANEMO,
-                                                         Element.ELECTRO,
-                                                         Element.DENDRO,
-                                                         Element.CRYO,
-                                                         Element.GEO, ]
+__all__ = [
+    "AbstractDices",
+    "ActualDices",
+    "Dices",
+]
 
 
 class Dices:
     """
     Base class for dices
     """
+
+    # higher PRIORITY - should be spent first
+    _ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY: tuple[Element, ...] = (
+        Element.PYRO,
+        Element.HYDRO,
+        Element.ANEMO,
+        Element.ELECTRO,
+        Element.DENDRO,
+        Element.CRYO,
+        Element.GEO,
+    )
+
+    _ELEMENTS_AND_OMNI_BY_DECREASING_GLOBAL_PRIORITY: tuple[Element, ...] = tuple(
+        list(_ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY) + [Element.OMNI]
+    )
+
+    _ELEMENTS_AND_OMNI = _ELEMENTS_AND_OMNI_BY_DECREASING_GLOBAL_PRIORITY
+
+    _ELEMENTS = _ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY
+
+    _NUMBER_OF_ELEMENTS: int = len(_ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY)
+
+    _ELEMENTS_BY_INCREASING_GLOBAL_PRIORITY: tuple[Element, ...] \
+        = _ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY[::-1]
+
+    _ELEMENTS_AND_OMNI_BY_INCREASING_GLOBAL_PRIORITY = \
+        _ELEMENTS_AND_OMNI_BY_DECREASING_GLOBAL_PRIORITY[::-1]
```

Comment:
These constants are only used by `ActualDices` so they should fall under `ActualDices` instead of `Dices`

Reply:
done. Also, I rewrited it so code now get all of them from one list (`_LEGAL_ELEMS_ORDERED`).